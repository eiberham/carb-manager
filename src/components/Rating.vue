<template>
  <section class="rating">
    <ul class="rating__score">
      <li
        class="star"
        :class="{
          'full-star': star.class == 'full-star',
          'half-star': star.class == 'half-star',
          'empty-star': star.class == 'empty-star'
        }"
        v-for="(star, index) in stars"
        :key="index"
      ></li>
    </ul>
    <div class="rating__count">{{ count }} ratings</div>
  </section>
</template>

<script>
export default {
  props: {
    count: {
      type: Number,
      required: true
    },
    score: {
      type: Number,
      required: false,
      default: 0
    }
  },
  computed: {
    stars() {
      return [...Array(5)].reduce((acc, _, index) => {
        const val = index + 1;
        if (val - this.score <= 0) {
          acc.push({ class: "full-star" });
        } else if (val - Math.abs(this.score) === 0.5) {
          acc.push({ class: "half-star" });
        } else {
          acc.push({ class: "empty-star" });
        }
        return acc;
      }, []);
    }
  }
};
</script>

<style scoped>
* {
  box-sizing: border-box;
}

.rating {
  display: flex;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  margin: 0;
  padding: 0;
}

.rating__score {
  display: inline;
  margin: 0;
  padding: 0;
  margin-right: 4px;
}

.star {
  background-repeat: no-repeat;
  background-size: 12px 12px;
  background-position: center top;
  width: 12px;
  height: 12px;
  list-style: none;
  float: left;
}

.star.full-star {
  background-image: url("../assets/star.svg");
}

.star.half-star {
  background-image: url("../assets/half-star.svg");
}

.star.empty-star {
  background-image: url("../assets/empty-star.svg");
}

.rating__count {
  font-family: Proxima Nova;
  font-style: normal;
  font-weight: 500;
  font-size: 14px;
  line-height: 14px;
  color: #1ca677;
  display: flex;
  justify-content: center;
  align-items: center;
  margin-left: 4px;
}
</style>
