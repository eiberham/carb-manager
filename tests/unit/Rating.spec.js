import { mount } from "@vue/test-utils";
import Rating from "../../src/components/Rating.vue";

describe("Rating.vue", () => {
  const props = {
    count: 200,
    score: 4.5
  };

  const wrapper = mount(Rating, {
    propsData: {
      ...props
    }
  });

  test("is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  test("renders stars correctly", () => {
    expect(wrapper.findAll(".star.full-star")).toHaveLength(4);
    expect(wrapper.findAll(".star.half-star")).toHaveLength(1);
  });

  test("matches snapshot", () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
