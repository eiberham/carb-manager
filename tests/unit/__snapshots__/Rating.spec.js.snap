// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating.vue matches snapshot 1`] = `
<section
  class="rating"
>
  <ul
    class="rating__score"
  >
    <li
      class="star full-star"
    />
    <li
      class="star full-star"
    />
    <li
      class="star full-star"
    />
    <li
      class="star full-star"
    />
    <li
      class="star half-star"
    />
  </ul>
   
  <div
    class="rating__count"
  >
    200 ratings
  </div>
</section>
`;
